### PR TITLE
Implement Splatoon Plug-in

### PR DESCRIPTION
**Pull request by DeathlyPlays**
This PR should be merged alongside #161 for the full effect (although even without it this will function properly)

### CODE REVIEW

File: wavelength-plugins/splatoon.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
...ershot Jr.", "Dapple Dualies", "Dapple Dualies", "Dualie Squelchers", "Dynamo Roller", "E-Liter 4K", "E-Liter 4K Scope", "Enperry Splat Dualies", "Firefin Splat Charger", "Firefine Splatterscope", "Flingza Roller", "Foil Flingza Roller", "Forge Splattershot Pro", "Glooga Dualies", "Gold Dynamo Roller", "Goo Tuber", "H-3 Nozzlenose", "Heavy Splatling", "Heavy Splatling Deco", "Hero Blaster Replica", "Hero Brella Replica", "Hero Charger Replica", "Hero Dualie Replica", "Hero Roller Replica", "Hero Shot Replica", "Hero Slosher Replica", "Hero Splatling Replica", "Herobrush Replica", "Hydra Splatling", "Inkbrush", "Inkbrush Nouveau", "Jet Squelcher", "Krak-On Splat Roller", "L-3 Nozzlenose", "L-3 Nozzlenose D", "Luna Blaster", "Mini Splatling", "N-ZAP '85", "N-ZAP '89", "Octobrush", "Octobrush Nouveau", "Range Blaster", "Rapid Blaster", "Rapid Blaster Deco", "Rapid Blaster Pro", "Slosher", "Slosher Deco", "Sloshing Machine", "Splash-o-matic", "Splat Brella", "Splat Charger", "Splat Dualies", "Splat Roller", "Splatterscope", "Splattershot", "Splattershot Jr.", "Splattershot Pro", "Sploosh-o-matic", "Squeezer", "Tenta Brella", "Tentatek Splattershot", "Tri-Slosher", "Undercover Brella"];
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: define the weapons globally in the file.

File: wavelength-plugins/splatoon.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	"spl2n": "splatoon",
+	"splatoon2": "splatoon",
+	splatoon: {
+		ranking: "rank",
+		ranks: "rank",
+		rank: {
+			clamblitz: "cb",
+			cb: function (target, room, user) {
+				if (!target) return this.parse("/splatoonhelp");
+				if (!["c-", "c", "c+", "b-", "b", "b+", "a-", "a", "a+", "s", "s+"].includes(target.toLowerCase())) return this.errorReply(`Invalid Ranking; check your spelling?`);
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: Use the proper case for a rank ex: `C-`, `B`, `S+`.
Also set the target to the proper case before testing so when its written to the DB its done right.
`target = target.trim().toUpperCase();`
Do this for all 4 commands.

File: wavelength-plugins/splatoon.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
+/****************************
+ * Splatoon Plug-in for PS	*
+ * Created by Insist		*
+ ****************************/
+
+"use strict";
+
+let SPLATFEST = false;
+let splatfestTeams = [];
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: Id combine this into one object such as
```javascript
let SPLATFEST = {alpha: null, bravo: null, active: false};
```
When setting up a splatfest, require that the two teams are passed to the command creating it.

File: wavelength-plugins/splatoon.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+			let targetUser = Users.get(target);
+			let username = (targetUser ? targetUser.name : target);
+			showProfile();
+
+			function IGN(user) {
+				if (!Db.splatign.has(user)) return ``;
+				return ` (<strong>IGN</strong>: ${Db.splatign.get(user)})`;
+			}
+
+			function showProfile() {
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: Dosent need to be a function.

File: wavelength-plugins/splatoon.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+			target = toId(target);
+			if (target.length > 18) return this.errorReply("Usernames cannot exceed 18 characters.");
+			let self = this;
+			let targetUser = Users.get(target);
+			let username = (targetUser ? targetUser.name : target);
+			showProfile();
+
+			function IGN(user) {
+				if (!Db.splatign.has(user)) return ``;
+				return ` (<strong>IGN</strong>: ${Db.splatign.get(user)})`;
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: Just run this check when building the profile.

File: wavelength-plugins/splatoon.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+			},
+		},
+
+		name: "ign",
+		ingamename: "ign",
+		ign: function (target, room, user) {
+			if (!this.canTalk()) return false;
+			if (!target || target.length > 10) return this.errorReply(`Your IGN must be between 1-10 characters long.`);
+			Db.splatign.set(user.userid, target);
+			return this.sendReply(`Your IGN has been set as: "${target}".`);
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: In-Game Name

File: wavelength-plugins/splatoon.js (hunk 1 above)
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: I tried that and for some reason it didn't work

**Comment by DeathlyPlays** [comment]
@HoeenCoder Honestly idk how to get each of the items to record to one Db (when setting it)

File: wavelength-plugins/splatoon.js (hunk 1 above)
Comment (HoeenCoder) [reply wavelength-plugins/splatoon.js]: It should work, can you show me what you did?

**Comment by HoeenCoder** [comment]
Do you know how an object works?

**Comment by DeathlyPlays** [comment]
Honestly I feel I do, but I've forgot exactly

File: wavelength-plugins/splatoon.js (hunk 1 above)
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: Gotcha, sadly I can't use the array for the setweapon thing it doesn't support arrays feelsbd

**Comment by HoeenCoder** [comment]
Ask me to explain them to you later

File: wavelength-plugins/splatoon.js [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+				SPLATFEST.bravo = null;
+			},
+
+			j: "join",
+			setteam: "join",
+			jointeam: "join",
+			join: function (target, room, user) {
+				if (!target) return this.parse(`/splatoonhelp`);
+				if (!SPLATFEST.active) return this.errorReply(`There is currently not a Splatfest. :(`);
+				if (!SPLATFEST.alpha.includes(toId(target)) || !SPLATFEST.bravo.includes(toId(target))) return this.errorReply(`This is not a Splatfest team.`);
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: Since your comparing exact strings, don't use includes. just compare them with `===`

Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: oooo ok

**Comment by DeathlyPlays** [comment]
switchfc is the only other Db added other than splatoon and that is intentional so I can call it in profiles as added in #161

**Comment by DeathlyPlays** [comment]
http://prntscr.com/ic0xjr finished product

File: wavelength-plugins/splatoon.js [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+				if (!SPLATFEST.active) return this.errorReply(`There is currently not a Splatfest. :(`);
+				return this.sendReplyBox(`<strong>Splatfest Teams:</strong> ${SPLATFEST.alpha} and ${SPLATFEST.bravo}`);
+			},
+		},
+
+		name: "ign",
+		ingamename: "ign",
+		ign: function (target, room, user) {
+			if (!this.canTalk()) return false;
+			if (!target || target.length > 10) return this.errorReply(`Your IGN must be between 1-10 characters long.`);
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]: add `!target.trim()` too

File: wavelength-plugins/splatoon.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+			if (!this.runBroadcast()) return;
+			if (!target) target = user.userid;
+			target = toId(target);
+			if (target.length > 18) return this.errorReply("Usernames cannot exceed 18 characters.");
+			let self = this;
+			let targetUser = Users.get(target);
+			let username = (targetUser ? targetUser.name : target);
+			let splatProfile = Db.splatoon.get(toId(username), {ranks: {}});
+
+			function IGN() {
```
[the remark this answers is not in the document]
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: The issue is having it on the same line as say names & level

File: wavelength-plugins/splatoon.js [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+			remove: "off",
+			off: function (target, room, user) {
+				if (!this.can("ban", null, room)) return this.errorReply(`You must be a Room Moderator or higher to use this command.`);
+				if (room.id !== "splatoon") return this.errorReply(`This command only works in the Splatoon room.`);
+				if (!SPLATFEST.active) return this.errorReply(`Splatfest is not currently active.`);
+				SPLATFEST.active = false;
+				let splatfestUsers = Db.splatoon.keys();
+				for (let u in splatfestUsers) {
+					let splatProfile = Db.splatoon.get(splatfestUsers[u], {ranks: {}});
+					delete splatProfile.splatfest;
```
[the remark this answers is not in the document]
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: Inside the for loop or before it?

File: wavelength-plugins/splatoon.js [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
...ynamo Roller", "Goo Tuber", "H-3 Nozzlenose", "Heavy Splatling", "Heavy Splatling Deco", "Hero Blaster Replica", "Hero Brella Replica", "Hero Charger Replica", "Hero Dualie Replica", "Hero Roller Replica", "Hero Shot Replica", "Hero Slosher Replica", "Hero Splatling Replica", "Herobrush Replica", "Hydra Splatling", "Inkbrush", "Inkbrush Nouveau", "Jet Squelcher", "Krak-On Splat Roller", "L-3 Nozzlenose", "L-3 Nozzlenose D", "Light Tetra Dualies", "Luna Blaster", "Luna Blaster Neo", "Mini Splatling", "Neo Splash-o-matic", "Neo Sploosh-o-matic", "New Squiffer", "N-ZAP '85", "N-ZAP '89", "Octobrush", "Octobrush Nouveau", "Range Blaster", "Rapid Blaster", "Rapid Blaster Deco", "Rapid Blaster Pro", "Slosher", "Slosher Deco", "Sloshing Machine", "Sloshing Machine Neo", "Sorella Brella", "Splash-o-matic", "Splat Brella", "Splat Charger", "Splat Dualies", "Splat Roller", "Splatterscope", "Splattershot", "Splattershot Jr.", "Splattershot Pro", "Sploosh-o-matic", "Squeezer", "Tenta Brella", "Tentatek Splattershot", "Tri-Slosher", "Tri-Slosher Nouveau" "Undercover Brella", "Zink Mini Splatling"].includes(target)) return this.errorReply(`Invalid weapon; check your spelling? [case sensitive]`);
```
Comment (HoeenCoder) [inline wavelength-plugins/splatoon.js]:
```javascript
if (!weapons.includes(target)) // ect
```

File: wavelength-plugins/splatoon.js [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+		splatfest: {
+			enable: "on",
+			start: "on",
+			new: "on",
+			make: "on",
+			on: function (target, room, user) {
+				if (!this.can("ban", null, room)) return this.errorReply(`You must be a Room Moderator or higher to use this command.`);
+				if (room.id !== "splatoon") return this.errorReply(`This command only works in the Splatoon room.`);
+				let targets = target.split(",");
```
Comment (Mystifi) [inline wavelength-plugins/splatoon.js]: `let [team1, team2] = target.split(',').map(p => p.trim());`

File: wavelength-plugins/splatoon.js [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+			end: "off",
+			cancel: "off",
+			remove: "off",
+			off: function (target, room, user) {
+				if (!this.can("ban", null, room)) return this.errorReply(`You must be a Room Moderator or higher to use this command.`);
+				if (room.id !== "splatoon") return this.errorReply(`This command only works in the Splatoon room.`);
+				if (!SPLATFEST.active) return this.errorReply(`Splatfest is not currently active.`);
+				SPLATFEST.active = false;
+				let splatfestUsers = Db.splatoon.keys();
+				for (let u in splatfestUsers) {
```
Comment (Mystifi) [inline wavelength-plugins/splatoon.js]:
```js
for (let u of splatfestusers) {
   let splatProfile = Db.splatoon.get(u, {ranks: {}});
   delete splatProfile.splatfest;
   Db.splatoon.set(u, splatProfile);
}
```

File: wavelength-plugins/splatoon.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mystifi) [reply wavelength-plugins/splatoon.js]: Before `delete splatProfile.splatfest;`.

File: wavelength-plugins/splatoon.js [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+			Db.splatoon.set(user.userid, splatProfile);
+			return this.sendReply(`Your Level has been set to: Level ${target}.`);
+		},
+
+		profile: function (target, room, user) {
+			if (!this.runBroadcast()) return;
+			if (!target) target = user.userid;
+			target = toId(target);
+			if (target.length > 18) return this.errorReply("Usernames cannot exceed 18 characters.");
+			let self = this;
```
Comment (Mystifi) [inline wavelength-plugins/splatoon.js]: No need for binding `this` to `self`; remove this variable declaration.

File: wavelength-plugins/splatoon.js [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+				return ` (<strong>IGN:</strong> ${splatProfile.ign})`;
+			}
+
+			function splatLevel() {
+				if (!splatProfile.level) return ``;
+				return ` <strong>Level:</strong> ${splatProfile.level}`;
+			}
+
+			let profile = ``;
+			profile += `<div><strong>Name:</strong> ${WL.nameColor(toId(username), true, true)}${IGN(toId(username))}${splatLevel(toId(username))}<br />`;
```
Comment (Mystifi) [inline wavelength-plugins/splatoon.js]: Consolidate this template string into the original declaration of `profile`.
Basically, just do:
```js
let profile = `<div><strong>[rest of L201...]`;
```

File: wavelength-plugins/splatoon.js [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+				if (!["C-", "C", "C+", "B-", "B", "B+", "A-", "A", "A+", "S", "S+"].includes(target)) return this.errorReply(`Invalid Ranking; check your spelling?`);
+				splatProfile.ranks.tc = target;
+				Db.splatoon.set(user.userid, splatProfile);
+				return this.sendReply(`You have successfully set your Tower Control ranking to "${target}".`);
+			},
+
+			salmonrun: "sr",
+			sr: function (target, room, user) {
+				if (!target) return this.parse("/splatoonhelp");
+				target = target.trim().split("-").map(x => { return x.substring(0, 1).toUpperCase() + x.substring(1).toLowerCase(); }).join("-");
```
Comment (Mystifi) [inline wavelength-plugins/splatoon.js]: `target = target.trim().split('-').map(x => x.charAt(0).toUpperCase() + x.slice(1).toLowerCase()).join('-');`

File: wavelength-plugins/splatoon.js [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
...", "E-Liter 4K Scope", "Enperry Splat Dualies", "Firefin Splat Charger", "Firefine Splatterscope", "Flingza Roller", "Foil Flingza Roller", "Forge Splattershot Pro", "Glooga Dualies", "Gold Dynamo Roller", "Goo Tuber", "H-3 Nozzlenose", "Heavy Splatling", "Heavy Splatling Deco", "Hero Blaster Replica", "Hero Brella Replica", "Hero Charger Replica", "Hero Dualie Replica", "Hero Roller Replica", "Hero Shot Replica", "Hero Slosher Replica", "Hero Splatling Replica", "Herobrush Replica", "Hydra Splatling", "Inkbrush", "Inkbrush Nouveau", "Jet Squelcher", "Krak-On Splat Roller", "L-3 Nozzlenose", "L-3 Nozzlenose D", "Luna Blaster", "Mini Splatling", "N-ZAP '85", "N-ZAP '89", "Octobrush", "Octobrush Nouveau", "Range Blaster", "Rapid Blaster", "Rapid Blaster Deco", "Rapid Blaster Pro", "Slosher", "Slosher Deco", "Sloshing Machine", "Splash-o-matic", "Splat Brella", "Splat Charger", "Splat Dualies", "Splat Roller", "Splatterscope", "Splattershot", "Splattershot Jr.", "Splattershot Pro", "Sploosh-o-matic", "Squeezer", "Tenta Brella", "Tentatek Splattershot", "Tri-Slosher", "Undercover Brella"].includes(target)) return this.errorReply(`Invalid weapon; check your spelling? [case sensitive]`);
```
[the remark this answers is not in the document]
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: I meant to use the global array but idk why it wasn't working before; now it does.

File: wavelength-plugins/splatoon.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: nvm

**Comment by DeathlyPlays** [comment]
TODO: 

- [ ] Ensure capitalization on /splatoon join is SPLATFEST.bravo/SPLATFEST.alpha
- [ ] Transfer splatLevel and IGN out of the functions and maintain being on the same line as Name:

File: wavelength-plugins/splatoon.js (hunk 11 above)
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: Done kek

File: wavelength-plugins/splatoon.js (hunk 14 above)
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: Oh true because its no longer a function

File: wavelength-plugins/splatoon.js (hunk 15 above)
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: I was debating whether or not to do that

File: wavelength-plugins/splatoon.js (hunk 13 above)
Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: gonna merge that idea w/ @HoeenCoder's to use an if statement before the delete

**Comment by DeathlyPlays** [comment]
P.S. the reason I put those on TODO is because I have literally zero clue how I would do it...
@HoeenCoder @Mystifi

File: wavelength-plugins/splatoon.js [hunk 18]
```diff
@@ -97,17 +97,16 @@ exports.commands = {
 			on: function (target, room, user) {
 				if (!this.can("ban", null, room)) return this.errorReply(`You must be a Room Moderator or higher to use this command.`);
 				if (room.id !== "splatoon") return this.errorReply(`This command only works in the Splatoon room.`);
-				let targets = target.split(",");
-				for (let u in targets) targets[u] = targets[u].trim();
-				if (!targets[1]) return this.parse("/splatoonhelp");
+				let [team1, team2] = target.split(",").map(p => p.trim());
+				if (!team2) return this.parse("/splatoonhelp");
```
Comment (Mystifi) [inline wavelength-plugins/splatoon.js]: `if (!(team1 && team2))` to ensure that both `team1` and `team2` are present and are truthy (not empty strings).

Comment (DeathlyPlays) [reply wavelength-plugins/splatoon.js]: oh

**Pull request merged**